### DISCUSSION TITLE
Installing on Linux / Ubuntu 22.04 (Pop OS)

### DISCUSSION THREAD

**Discussion by Xalalau** [WilliamVenner/gmpublisher#178] [General]
I managed to compile gmpublisher and I'd like to share this feat with you!

So far I had no crashes and my mouse input is fine (https://github.com/WilliamVenner/gmpublisher/discussions/102) and everything is working as expected :)

![image](https://github.com/WilliamVenner/gmpublisher/assets/5098527/76da35d9-7a42-4132-b274-e556917fed2f)

<details>
<summary>More screenshots</summary><p>

![image](https://github.com/WilliamVenner/gmpublisher/assets/5098527/c733663d-b6f4-4f3f-bef2-01f107f8f357)

![image](https://github.com/WilliamVenner/gmpublisher/assets/5098527/2cda0ab4-d792-4317-b30c-fee52dfcedc5)

</p></details>

Here are my compilation steps:

<details>
<summary> v2.9.2 (tested on Pop OS 22.04)</summary><p>

Compile in a VM or Container to keep your system clean.

```sh
# Install Tauri prerequisites (linux package requirements + Rust 1.70.0)
# https://tauri.app/v1/guides/getting-started/prerequisites

# Install nodejs v20.3.1 https://nodejs.org/en/download/package-manager

# Download steam sdk version 151 https://partner.steamgames.com/doc/sdk
# unzip it into your Downloads folder

sudo apt install libgtksourceview-3.0-dev clang

npm i -g corepack

corepack prepare yarn@stable --activate

sudo apt install git

git clone https://github.com/WilliamVenner/gmpublisher.git

cd gmpublisher

git checkout tags/2.9.2

npm i -f # Force to install fsevents

yarn install

npm install 

npm run build

cd src-tauri/

# in Cargo.toml add this line after [dependencies]:
openssl = { version = "0.10", features = ["vendored"] }

STEAM_SDK_LOCATION=~/Downloads/sdk cargo build --release

cd target/release

# Install gmpublisher in your system:
sudo cp gmpublisher /usr/bin
sudo cp libsteam_api.so /usr/lib

# Execute it:
gmpublisher
```

</p></details>

<details>
<summary>v2.11.1 (tested on Linux Mint 22.2) (broken)</summary><p>

When trying to compile v2.11.1 on Mint 22.2 I learned that the code is stuck on Tauri version 1.6, which makes it dependent on older versions of libwebkit2gtk (4.0, while Mint uses 4.1). This alone causes several problems in the compilation and I simply break the system to make it work. Anyway, the executable still has conflicting dependencies when used in clean systems, so it's kinda pointless. I guess the ideal solution here would be to continue compiling gmpublisher on Ubuntu 22.04 and make a appimage of it, but I don't know how to do that. Below is my experiment, if anyone is interested.

Note: This MUST be done in a VM or Container!!!!! I really break the system!!!

v2.11.1 (tested on Mint 22.2):
```sh
# Start with
sudo apt install pkg-config libsoup2.4-dev

# Then remove the audio
sudo apt remove pulseaudio pipewire-pulse libcanberra-pulse

# Destroy the system to get rid of libsoup-3.0-0
sudo apt remove libsoup-3.0-*

# Get libwebkit2gtk-4.0 and libwebkit2gtk-4.0-dev with some dependencies using this gr8 gambiarra
sudo apt remove libwebkit2gtk-4.1-*
sudo apt autoremove
echo "deb http://archive.ubuntu.com/ubuntu jammy main universe" | sudo tee /etc/apt/sources.list.d/jammy-temp.list
sudo apt update
sudo apt install libwebkit2gtk-4.0-dev libwebkit2gtk-4.0-37
sudo rm /etc/apt/sources.list.d/jammy-temp.list
sudo apt update

# Install Tauri prerequisites (linux package requirements + Rust 1.91.1)
# https://tauri.app/v1/guides/getting-started/prerequisites

# Install nodejs v24.11.1 https://nodejs.org/en/download/package-manager

# Download steam sdk version 1.58a https://partner.steamgames.com/doc/sdk
# unzip it into your Downloads folder

sudo apt install libgtksourceview-3.0-dev clang

npm i -g corepack

corepack prepare yarn@stable --activate

sudo apt install git

git clone https://github.com/WilliamVenner/gmpublisher.git

cd gmpublisher

git checkout tags/2.11.1

npm i -f # Force to install fsevents

yarn install

npm install 

npm run build

cd src-tauri/

STEAM_SDK_LOCATION=~/Downloads/sdk cargo build --release

cd target/release

# Install gmpublisher in your system:
sudo cp gmpublisher /usr/bin
sudo cp libsteam_api.so /usr/lib

# Execute it:
gmpublisher
```

</p></details>

### Extra 1 - Finding the correct Steam SDK version

<details>

<summary>Locations</summary><p>

Just so you know, you search for steamworks in this file:
https://github.com/WilliamVenner/gmpublisher/blob/master/src-tauri/Cargo.toml

And find the correct SDK version here:
https://docs.rs/steamworks/latest/steamworks/

</p></details>

### Extra 2 - Download my binary (64 bits) & how to run it

[gmpublisher.zip](https://github.com/WilliamVenner/gmpublisher/files/11828525/gmpublisher.zip) (v2.9.2 for 22.04)
(test the binary links with ldd, if needed)

<details>

<summary>Installation</summary><p>

To run my compiled gmpublisher you need to unzip the files into ~/Applications/gmpublisher and execute this command:
```sh
cd ~/Applications/gmpublisher
LD_LIBRARY_PATH=~/Applications/gmpublisher ./gmpublisher
```

You can also add it to the system:
```sh
cd ~/Applications/gmpublisher
sudo cp gmpublisher /usr/bin
sudo cp libsteam_api.so /usr/lib
gmpublisher
```

</p></details>

### Extra 3 - Create a system shortcut

<details>

<summary>Set it up</summary><p>

If you're using a distro link Linux Mint, you can create the shortcut with a tool like this one:

![image](https://github.com/WilliamVenner/gmpublisher/assets/5098527/f59b0a40-00cf-4f65-b102-e324012d9138)

On Pop OS, you can do it manually with these steps:

Save this icon:

![gmpublisher](https://github.com/WilliamVenner/gmpublisher/assets/5098527/81a60996-9fc2-4a66-8725-60b063e72687)
 
as ``~/.local/share/icons/hicolor/256x256/apps/gmpublisher.png``

Or just go ahead and create the file manually:

1) Copy this text (change the YOURUSERNAME value) (I'm assuming you installed gmpublisher in the system):
```
[Desktop Entry]
Type=Application
Name=gmpublisher
Exec=/user/bin/gmpublisher
Icon=/home/YOURUSERNAME/.local/share/icons/hicolor/256x256/apps/gmpublisher.png
TryExec=gmpublisher
Comment=A powerful and feature-packed Workshop publisher for Garry's Mod!
```

2) Paste it into ``~/.local/share/applications/gmpublisher.desktop`` (create this file)

3) Run this command:

``sudo chmod +x ~/.local/share/applications/gmpublisher.desktop``

Ta-da!

![image](https://github.com/WilliamVenner/gmpublisher/assets/5098527/a36ac4be-348b-4ff3-b9b3-b7ae3d031e76)
(restart your session if the icon doesn't appear)

</p></details>

I feel like Linux is better covered now.

**Comment by MrGenie151** [comment]
Thank you so much! The gmpublisher you compiled did manage to open on my Linux Mint 21.1 PC, so it should be good for most other people (hopefully).

**Comment by Def-Try** [comment]
I think you forgot to change Icon value in gmpublisher.desktop example. Anyways, thank you very much! I finally was able to make github workflow for building.

**Reply by Xalalau** [reply]
Fixed*

**Comment by Xalalau** [comment]
Ugly workaround for Ubuntu 24.04 based distros (tested on Linux Mint):

```bash
sudo ln -sf /usr/lib/x86_64-linux-gnu/libwebkit2gtk-4.1.so.0 /usr/lib/x86_64-linux-gnu/libwebkit2gtk-4.0.so.37
sudo ln -sf /usr/lib/x86_64-linux-gnu/libjavascriptcoregtk-4.1.so.0 /usr/lib/x86_64-linux-gnu/libjavascriptcoregtk-4.0.so.18
```

**Comment by davidmacalister** [comment]
nice!

**Comment by RubberDev** [comment]
Don't know if I did something wrong but it didn't work for me lol

**Comment by awsifying** [comment]
Ty man, working on endeveaurOS.

**Comment by WilliamVenner** [comment]
#333
